Ap 2304 Backend service for sharing artifact (Log, Process, Folder)

franklma opened the pull request:
Add sharing APIs in WorkspaceService, which basically a mapper between ACL access type and flags in group_* tables.

Commit:
Fix metadata permission after deleting of Log

Commit:
Fix metadata permission after deleting of Log

Commit:
Remove unnecessary cascading in GroupUserMetadata

Commit:
Remove unnecessary cascading in GroupUserMetadata

Commit:
Feature: CRUD functions for ACL of Log/Process/Folder

Commit:
Feature: CRUD functions for ACL of Log/Process/Folder

raboczi committed:
Re-enabled pre-existing Checkstyle code. (AP-2409)

Updated Checkstyle from v2 to v3; the old (partially incompatible checkstyle XML is retained as checkstyle-v2.xml).
The v3 (checkstyle.xml) is mostly default; Javadoc, DesignForExtension, and OperatorWrap are notable changes.

Commit:
Add unit tests

Commit:
Add unit tests

Commit:
Fix equal and hashcode methods in Usermetadata POJO

Commit:
Fix equal and hashcode methods in Usermetadata POJO

Commit:
comment

Commit:
comment

nolantellis commented on Apromore-Core-Components/Apromore-Manager/src/main/java/org/apromore/service/impl/UserMetadataServiceImpl.java:
```diff
@@ -141,6 +142,42 @@ public void saveUserMetadata(String userMetadataContent, UserMetadataTypeEnum us
+        Set<UsermetadataLog> usermetadataLogSet =
+                new HashSet<>(usermetadataLogRepo.findByLog(logRepo.findUniqueByID(logId)));
+
+        if (usermetadataLogSet.size() != 0) {
```
if loop not required.

nolantellis commented on Apromore-Core-Components/Apromore-Manager/src/main/java/org/apromore/service/impl/UserMetadataServiceImpl.java:
```diff
@@ -141,6 +142,42 @@ public void saveUserMetadata(String userMetadataContent, UserMetadataTypeEnum us
+        if (usermetadataLogSet.size() != 0) {
+
+            for (UsermetadataLog usermetadataLog : usermetadataLogSet) {
+                Usermetadata u = usermetadataLog.getUsermetadata();
```
Not important now, but a quick way to do this is get all the Ids in 1 db quiery, and then use in query to get all list of groupmeta data , and then so the needful.

brucenguyen1 replied:
What happens if an arbitrary groupRowGuid value is passed to the findByLogAndGroup() method and a null is returned at line 154? This kind of corner case can be checked in other places and would be great to be in unit tests to be solid.

franklma replied:
Will add unit tests to cover these corner cases.

nolantellis commented on Apromore-Core-Components/Apromore-Manager/src/main/java/org/apromore/service/impl/UserMetadataServiceImpl.java:
```diff
@@ -182,28 +219,18 @@ public void saveUserMetadataPermissions(Integer logId, String groupRowGuid, bool
-                        groupUsermetadataRepo.delete(g);
-                    }
-                }
+            if(g.getUsermetadata().getGroupUserMetadata().size() == 1) { // if this metadata will be an orphan
```
Not sure, but you may get nullpointer over here if getUserMetadata is null

nolantellis commented on Apromore-Core-Components/Apromore-Manager/src/main/java/org/apromore/service/impl/WorkspaceServiceImpl.java:
```diff
@@ -695,26 +697,125 @@ private void removeGroupLog(Group group, Log log) {
+        Map<Group, AccessType> groupAccessTypeMap = new HashMap<>();
+
+        for (GroupLog g : getGroupLogs(logId)) {
+            if (g.getHasRead() && g.getHasWrite() && g.getHasOwnership()) {
```
you can group. (g.getHasRead() && g.getHasWrite() && g.getHasOwnership()) inside GroupLog with helper method isOwner() or anyother name. and make this 1 line ternary operator

nolantellis commented on Apromore-Core-Components/Apromore-Manager/src/main/java/org/apromore/service/impl/WorkspaceServiceImpl.java:
```diff
@@ -695,26 +697,125 @@ private void removeGroupLog(Group group, Log log) {
+        Map<Group, AccessType> groupAccessTypeMap = new HashMap<>();
+
+        for (GroupProcess g : getGroupProcesses(processId)) {
+            if (g.getHasRead() && g.getHasWrite() && g.getHasOwnership()) {
```
Repeated code

franklma replied:
Are there ways to reduce repeating code here other than abstract comparing and use ternary operator? Thanks

brucenguyen1 replied:
This block of code: {if...then...else...} with hasRead, hasWrite, hasOwnership need a higher level of abstraction to minimize various miscellaneous changes to source code in many places like this if these any rules to hasRead, hasWrite, etc. change (which always happens). From the repeated code block, higher abstraction methods are needed to hide these details: e.g. checkHasEditorAccessType(), checkHasViewerAccessType, checkHasOwnerAccessType. Again, these authorization checks should be responsibilty of a separate class.

brucenguyen1 replied:
Alternatively, you can create just one method called checkPermissionByAccessType(AccessType accessType, Permission permission). Note that a new class called Permission is introduced here to wrap the details of hasRead, hasOwner, hasXXX. It makes the code much cleaner. Btw, as you're adding code relating to these permissions, I suggest we need a little refactoring here: adding the Permission class and replacing all references to details of hasXXX to this Permission class.

nolantellis reviewed (changes requested):
Things look alright. If you want to make the changes suggested , let me know, I may help in some changes. Else to get this work done quickly ,your free to merge.

brucenguyen1 commented on Apromore-Core-Components/Apromore-Manager/src/main/java/org/apromore/util/AccessType.java:
```diff
@@ -0,0 +1,50 @@
+        this.flagId = flagId;
+    }
+
+    public Integer getFlogId() {
+        return flagId;
+    }
+}
```
Misspelling: getFlagId.

brucenguyen1 replied:
flagId can be an internal variable name, but the method name getFlagId() doesn't show its meaning. It can be given a better name according its intention here: e.g. getAccessTypeCode? I would suggest to rename the flagId variable to be a better name like accessTypeCode.

brucenguyen1 commented on Apromore-Core-Components/Apromore-Manager/src/main/java/org/apromore/util/AccessType.java:
```diff
@@ -0,0 +1,50 @@
+
+    private final Integer flagId;
+
+    AccessType(Integer flagId) {
+        this.flagId = flagId;
+    }
```
Missing access modifier.

brucenguyen1 replied:
What happens if flagId is an arbitrary integer, e.g. 100?

brucenguyen1 commented on Apromore-Core-Components/Apromore-Manager/src/main/java/org/apromore/service/WorkspaceService.java:
```diff
@@ -174,4 +173,75 @@
+     * @param logId Log ID
+     * @return list of Group and AccessType pair
+     */
+    Map<Group, AccessType> getLogACL(Integer logId);
+
+    /**
+     * Get list of Group and AccessType pair of Log
+     * @param processId Process ID
+     * @return list of Group and AccessType pair
+     */
+    Map<Group, AccessType> getProcessACL(Integer processId);
+
+    /**
+     * Get list of Group and AccessType pair of Log
+     * @param folderId Folder ID
+     * @return list of Group and AccessType pair
+     */
+    Map<Group, AccessType> getFolderACL(Integer folderId);
+
+    /**
```
These XXXPermissions and XXXACL should be put in a separate class which is responsible for security/access right management. The WorkspaceService is responsible for managing the workspace such as create/remove/move folder, get list of subfolders, breadcrumbs, etc. The general principle is "one class one responsiblity". Access rights/permissions belong to a different type of responsiblity. The reason we need to care about responsibility in class design is because the class would be modified when its responsiblity changes, and thus, keep responsiblities in separate classes would help keep the code clean and minimal impact on the code when changes happen (always happen). Note that there is already a service class called SecurityService with a mix of authentication (username, password) and authorization (roles) - it's not clean, suggest to move XXXPermissions and XXXACL to a class called something like AuthorizationService, PermissionService, or AccessRightService.

franklma replied:
Yes. AuthorizationService class sounds a good place to contains access control methods, and we can move all the authorization (roles) methods from SecurityService to this class in the next step.

brucenguyen1 commented on Apromore-Core-Components/Apromore-Manager/src/test/java/org/apromore/service/impl/WorkspaceServiceImplTest.java:
```diff
@@ -362,5 +353,122 @@ public void testMoveFolder() throws Exception {
     }
-    
+
+    @Test
+    public void getLogACL() {
+
+        // Set up test data
+        Group group1 = createGroup(1, Group.Type.GROUP);
+        Group group2 = createGroup(2, Group.Type.GROUP);
+        Group group3 = createGroup(3, Group.Type.USER);
+        Group group4 = createGroup(4, Group.Type.USER);
+
+        Role role = createRole(createSet(createPermission()));
+        User user = createUser("userName1", group1, createSet(group1), createSet(role));
+
+        Workspace wp = createWorkspace(user);
+        Log log = createLog(user, createFolder("testFolder", null, wp));
+
```
Unit test methods should be named as testXXX, e.g. testGetLogACL, or test_getLogACL, testGetProcessACL, etc. This is to avoid miscommunication when referring to a test method: not to mislead it with the method under test.

franklma replied:
RIght. They were auto-generated by IntelliJ.

brucenguyen1 commented on Apromore-Core-Components/Apromore-Manager/src/main/java/org/apromore/service/impl/WorkspaceServiceImpl.java:
```diff
@@ -695,26 +697,125 @@ private void removeGroupLog(Group group, Log log) {
-//            groupLogRepo.delete(groupLog);
-//        }
-//    }
+    public Map<Group, AccessType> getLogACL(Integer logId) {
+
+        Map<Group, AccessType> groupAccessTypeMap = new HashMap<>();
+
+        for (GroupLog g : getGroupLogs(logId)) {
+            if (g.getHasRead() && g.getHasWrite() && g.getHasOwnership()) {
```
This method should be renamed as: getLogAccessType() as shown from its implementation, the return is AccessType. ACL is Access Control List which is not at this level. Same for other method: getFolderACL, getXXXACL.

franklma replied:
Yes. getLogAccessType is definitely more accurate.

brucenguyen1 reviewed (commented):
I have a number of comments. Some are related to a little bit refactoring which is needed to improve code while the new code is added.

brucenguyen1 commented on Apromore-Core-Components/Apromore-Manager/src/main/java/org/apromore/service/impl/WorkspaceServiceImpl.java:
```diff
@@ -695,26 +697,125 @@ private void removeGroupLog(Group group, Log log) {
+        }
+    }
+
+    // Delete Log's access right may lead to logical deleting of user metadata, which need username to fill UpdateBy
+    // field
+    public void deleteLogACL(Integer logId, String groupRowGuid, String username) throws UserNotFoundException {
+
+        removeLogPermissions(logId, groupRowGuid, username);
+    }
+
+    public void deleteProcessACL(Integer processId, String groupRowGuid) {
+
+        removeProcessPermissions(processId, groupRowGuid);
+    }
+
+    public void deleteFolderACL(Integer folderId, String groupRowGuid) {
+
+        removeFolderPermissions(folderId, groupRowGuid);
+    }
```
Relating to one comment above, these methods deleteXXXACL with a one-to-one mapping to removeXXXPermissions show that ACL = Permission. That's why the above comment suggests using a better name for getXXXACL (they should be getXXXAccessType instead), and ACL corresponds to Permission, so suggest using one consistent name in source code: either ACL or Permission, not both. Once consistent notion and name have been determined, these one-to-one mapping methods will not be needed any more.

Commit:
Merge remote-tracking branch 'origin/development' into AP-2304

Commit:
Minor refactoring and test correction

Commit:
Minor refactoring and test correction

raboczi committed:
Incorporated feedback to checkstyle.xml

- trailing whitespace is permitted if the line is entirely whitespace; this allows Eclipse's auto-indentation
- documentation links normalized to checkstyle.org

raboczi committed:
Merge pull request #507 from apromore/feature/AP-2409-enable-checkstyle

Re-enabled pre-existing Checkstyle code. (AP-2409)

Commit:
Add preConditions in liquidbase changeSet to cover all scenarios

Commit:
Add tableName in indexExists preCondition and change indexName to follow naming conventions

franklma committed:
Merge pull request #517 from apromore/AP-2419

Ap 2419 Add preConditions in LiquidBase changeSet to cover all scenarios

dreamind committed:
AP-2311 Increase notification hide delay

dreamind committed:
Merge pull request #518 from apromore/feature/AP-2311-2

AP-2311 Increase notification hide delay

Commit:
Refactor access control APIs out of WorkspaceService to AuthorizationService

Commit:
Reverse IN clause code due to possible incompatibility of spring JPA

Commit:
Fix regressions

Commit:
Rebase

Commit:
Change method signature

brucenguyen1 reviewed (approved):
As discussed, refactoring will need to do later in terms of method naming (mixed up feature access with resource access), method signature, and separation of code  to classes.
